pkg/pmem-csi-driver/parameters.go

```diff
@@ -0,0 +1,272 @@
+/*
+Copyright 2019 Intel Corporation
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package pmemcsidriver
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/docker/go-units"
+)
+
+type persistencyModel string
+type parameterOrigin int
+
+// Beware of API and backwards-compatibility breaking when changing these string constants!
+const (
+	parameterCacheSize        = "cacheSize"
+	parameterEraseAfter       = "eraseafter"
+	parameterName             = "name"
+	parameterPersistencyModel = "persistencyModel"
+	parameterVolumeID         = "_id"
+	parameterSize             = "size"
+
+	// Kubernetes v1.16+ adds this key to NodePublishRequest.VolumeContext
+	// while provisioning ephemeral volume.
+	parameterEphemeral = "csi.storage.k8s.io/ephemeral"
+
+	// Additional, unknown parameters that are okay.
+	parameterPodInfoPrefix = "csi.storage.k8s.io/"
+
+	persistencyNormal    persistencyModel = "normal" // In releases <= 0.6.x this was called "none", but not documented.
+	persistencyCache     persistencyModel = "cache"
+	persistencyEphemeral persistencyModel = "ephemeral" // only used internally
+
+	// Parameters are from storage class in controller CreateVolume.
+	createVolumeParameters parameterOrigin = iota
+	// Parameters are from master in node CreateVolume.
+	createVolumeInternalParameters
+	// Parameters are for an ephemeral volume in NodePublishVolume.
+	ephemeralVolumeParameters
+	// Parameters are for a persistent volume in NodePublishVolume.
+	persistentVolumeParameters
+	// Parameters as stored in node volume list.
+	nodeVolumeParameters
+)
+
+// validParameters is a whitelist of which parameters are valid in which context.
+var validParameters = map[parameterOrigin][]string{
+	// Parameters from Kubernetes and users for a persistent volume.
+	createVolumeParameters: []string{
+		parameterCacheSize,
+		parameterEraseAfter,
+		parameterPersistencyModel,
+		parameterPodInfoPrefix,
+	},
+
+	// These parameters are prepared by the master controller.
+	createVolumeInternalParameters: []string{
+		parameterCacheSize,
+		parameterEraseAfter,
+		parameterPersistencyModel,
+
+		parameterVolumeID,
+	},
+
+	// Parameters from Kubernetes and users.
+	ephemeralVolumeParameters: []string{
+		parameterEraseAfter,
+		parameterPodInfoPrefix,
+		parameterSize,
+	},
+
+	// The volume context prepared by CreateVolume.
+	persistentVolumeParameters: []string{
+		parameterPersistencyModel,
+		parameterCacheSize,
+		parameterEraseAfter,
+	},
+
+	// Internally we store everything except the volume ID
+	// which is handled separately.
+	nodeVolumeParameters: []string{
```

Comment:
It can be removed if we agree with https://github.com/intel/pmem-csi/pull/504/files#r362738437